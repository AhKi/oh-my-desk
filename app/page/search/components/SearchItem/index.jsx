import React from 'react';
import PropTypes from 'prop-types';
import ToggleButton from 'page/Components/ToggleButton';
import deleteIcon from 'assets/icon/icon-delete.svg';
import starIcon from 'assets/icon/icon-more.svg';
import './SearchItem.scss';

const propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    url: PropTypes.string,
  }),
};
const defaultProps = {
  item: {},
};

class SearchItem extends React.Component {
  render() {
    const { item } = this.props;

    return (
      <li className="SearchItem">
        <ToggleButton
          isCheck={item.isOpen}
        />
        <p className="SearchItem__Text">
          <span><strong>{item.name}</strong></span>
          <span>{item.url}</span>
        </p>
        <p className="SearchItem__BtnSet">
          <button type="button">
            <img src={starIcon} alt="" />
          </button>
          <button type="button">
            <img src={deleteIcon} alt="" />
          </button>
        </p>
      </li>
    );
  }
}

SearchItem.propTypes = propTypes;
SearchItem.defaultProps = defaultProps;

export default SearchItem;
