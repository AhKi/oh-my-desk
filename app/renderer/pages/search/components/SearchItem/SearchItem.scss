.SearchItem {
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 50px;
  border-bottom: 1px solid $color-cool-gray300;
  margin-bottom: -1px;

  &:hover {
    background-color: #EBEFF1;
  }
}


.SearchItem__select {
  background: linear-gradient(90deg, #9273EA, #7978FD,#6CA2FE);
  color: $color-white;

  .SearchItem__url-text {
    color: $color-white;
  }

  &:hover{
    background: linear-gradient(90deg, #7751E6, #5658FF,#488CFF);
  }
}


.SearchItem__namebox {
  display: flex;
  align-items: center;
  justify-content: left;
}


.SearchItem__toggle-btn {
  margin: 0 12px;
}


.SearchItem__Btn {
  color: inherit;
  cursor: default;
  display: flex;
  flex-direction: column;
  width: inherit;

}

.SearchItem__BtnSet {
  margin: 6px;
  display: flex;
  justify-content: center;
  align-items: center;
  white-space: nowrap;
  text-overflow: ellipsis;
  position: relative;
}

.SearchItem__title-text,
.SearchItem__url-text {
  max-width: 200px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.SearchItem__url-text {
  color: $color-middle-gray-blue;
}

.SearchItem__bookmark-btn {
  &:hover {
    background: $color-black-opacity15;
    border-radius: 4px;
  }

}

.SearchItem__more-btn {
  &:hover {
    background: $color-black-opacity15;
    border-radius: 4px;
  }
  svg {
    fill: #ffffff;
  }
}

.SearchItem__icon {
  width: 100%;
  height: 100%;
  opacity: 0.3;
}

.SearchItem__icon--select {
  opacity: 1;
}

//----moreBtnSscss----//
.SearchMoreMenu {
  position: absolute;
  width: 120px;
  top: 30px;
  right: 8px;
  background-color: $color-white;
  border: 1px solid rgb(216, 216, 216);
  box-shadow: 2px 3px 7px -2px rgba(59, 60, 67, .23);
  z-index: 1;

  li:hover {
    background-color: #E9EBEC;
  }
}

.SearchMoreMenu__btnSet {
  border-bottom: 1px solid rgb(216, 216, 216);
}

.SearchMoreMenu__btn {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px;
  width: 100%;


}