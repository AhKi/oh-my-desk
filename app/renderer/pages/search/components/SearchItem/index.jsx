import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import ToggleButton from 'renderer/components/ToggleButton';
import starIcon from 'assets/icon/icon-star-black.svg';
import HighlightParagraph from 'renderer/components/HighlightParagraph';
import './SearchItem.scss';

const propTypes = {
  isSelect: PropTypes.bool,
  item: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    url: PropTypes.string,
    searched: PropTypes.string,
  }),
  keyword: PropTypes.string,
  onCloseWidget: PropTypes.func,
  onShowWidget: PropTypes.func,
  onUpdateInfo: PropTypes.func,
};
const defaultProps = {
  isSelect: false,
  item: {},
  keyword: '',
  onCloseWidget() {},
  onShowWidget() {},
  onUpdateInfo() {},
};

class SearchItem extends React.Component {
  constructor(props) {
    super(props);
    this.handleFocusWidget = this.handleFocusWidget.bind(this);
    this.handleToggleWidget = this.handleToggleWidget.bind(this);
    this.handleToggleFavorites = this.handleToggleFavorites.bind(this);
    this.handleOpenWidget = this.handleOpenWidget.bind(this);
  }

  handleFocusWidget(e) {
    const { item, onShowWidget } = this.props;

    if (item.isOpen) {
      e.preventDefault();
      onShowWidget(item.id);
    }
  }

  handleToggleWidget() {
    const { item, onCloseWidget, onShowWidget } = this.props;

    if (item.isOpen) {
      onCloseWidget(item.id);
    } else {
      onShowWidget(item.id);
    }
  }

  handleOpenWidget() {
    const { item, onShowWidget } = this.props;

    onShowWidget(item.id);
  }

  handleToggleFavorites() {
    const { item, onUpdateInfo } = this.props;

    onUpdateInfo(item.id, { favorites: !item.favorites });
  }

  render() {
    const { keyword, item, isSelect } = this.props;
    const SearchItemClassName = cx('SearchItem', {
      SearchItem__select: isSelect,
    });
    const BookmarkClassName = cx('SearchItem__icon', {
      'SearchItem__icon--select': item.favorites,
    });

    return (
      <li className={SearchItemClassName}>
        <div className="SearchItem__toggle-btn">
          <ToggleButton
            isCheck={item.isOpen}
            onToggle={this.handleToggleWidget}
          />
        </div>
        <p className="SearchItem__Text">
          <button
            className="SearchItem__Btn"
            type="button"
            onClick={this.handleFocusWidget}
            onDoubleClick={this.handleOpenWidget}
          >
            <strong className="SearchItem__title-text">
              {(item.searched === 'both' || item.searched === 'name') ? (
                <HighlightParagraph
                  keyword={keyword}
                  content={item.name}
                />
              ) : <span>{item.name}</span>}
            </strong>
            <span className="SearchItem__url-text">
              {(item.searched === 'both' || item.searched === 'url') ? (
                <HighlightParagraph
                  keyword={keyword}
                  content={item.url}
                />
              ) : <span>{item.url}</span>}
            </span>
          </button>
        </p>
        <p className="SearchItem__BtnSet">
          <button
            className="SearchItem__bookmark-btn"
            type="button"
            onClick={this.handleToggleFavorites}
          >
            <img className={BookmarkClassName} src={starIcon} alt="" />
          </button>
        </p>
      </li>
    );
  }
}

SearchItem.propTypes = propTypes;
SearchItem.defaultProps = defaultProps;

export default SearchItem;
