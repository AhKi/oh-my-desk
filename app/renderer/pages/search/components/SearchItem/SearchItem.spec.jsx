import React from 'react';
import { shallow } from 'enzyme';
import SearchItem from '.';

describe('Test SearchItem Component', () => {
  const mockItem = {
    id: 'mock-id',
    name: 'mock-name',
    url: 'mock-url',
    isOpen: false,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match to snapshot', () => {
    const wrapper = shallow(<SearchItem item={mockItem} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should match to snapshot when item.searched === both', () => {
    const wrapper = shallow(
      <SearchItem
        item={{
          ...mockItem,
          searched: 'both',
        }}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('should call handleToggleWidget', () => {
    const onCloseWidget = jest.fn();
    const onShowWidget = jest.fn();

    const wrapper = shallow(
      <SearchItem
        item={mockItem}
        onCloseWidget={onCloseWidget}
        onShowWidget={onShowWidget}
      />,
    );

    it('when props.item.isOpen === false ', () => {
      expect(wrapper.instance().props.item.isOpen).toBe(false);

      wrapper.instance().handleToggleWidget();

      expect(onShowWidget).toHaveBeenCalledTimes(1);
      expect(onShowWidget).toHaveBeenCalledWith('mock-id');
      expect(onCloseWidget).toHaveBeenCalledTimes(0);
    });

    it('when props.item.isOpen === false ', () => {
      wrapper.setProps({ item: { ...mockItem, isOpen: true } });
      expect(wrapper.instance().props.item.isOpen).toBe(true);

      wrapper.instance().handleToggleWidget();

      expect(onCloseWidget).toHaveBeenCalledTimes(1);
      expect(onCloseWidget).toHaveBeenCalledWith('mock-id');
      expect(onShowWidget).toHaveBeenCalledTimes(0);
    });
  });

  describe('test handleFocusWidget', () => {
    it('when props.item.isOpen === true', () => {
      const onShowWidget = jest.fn();
      const mockEvent = {
        preventDefault: jest.fn(),
      };
      const wrapper = shallow(
        <SearchItem
          item={{
            ...mockItem,
            isOpen: true,
          }}
          onShowWidget={onShowWidget}
        />,
      );

      wrapper.instance().handleFocusWidget(mockEvent);

      expect(mockEvent.preventDefault).toHaveBeenCalledTimes(1);
      expect(onShowWidget).toHaveBeenCalledTimes(1);
      expect(onShowWidget).toHaveBeenCalledWith('mock-id');
    });

    it('when props.item.isOpen === false', () => {
      const onShowWidget = jest.fn();
      const mockEvent = {
        preventDefault: jest.fn(),
      };
      const wrapper = shallow(
        <SearchItem
          item={{
            ...mockItem,
            isOpen: false,
          }}
          onShowWidget={onShowWidget}
        />,
      );

      wrapper.instance().handleFocusWidget(mockEvent);
      expect(mockEvent.preventDefault).toHaveBeenCalledTimes(0);
      expect(onShowWidget).toHaveBeenCalledTimes(0);
    });
  });

  it('should call handleOpenWidget', () => {
    const onShowWidget = jest.fn();
    const wrapper = shallow(
      <SearchItem
        item={{
          ...mockItem,
          isOpen: false,
        }}
        onShowWidget={onShowWidget}
      />,
    );

    wrapper.instance().handleOpenWidget();

    expect(onShowWidget).toHaveBeenCalledTimes(1);
  });

  it('should call handleToggleFavorites', () => {
    const onUpdateInfo = jest.fn();
    const wrapper = shallow(
      <SearchItem
        item={{
          ...mockItem,
          favorites: false,
        }}
        onUpdateInfo={onUpdateInfo}
      />,
    );

    expect(wrapper.instance().props.item.favorites).toBe(false);
    wrapper.instance().handleToggleFavorites();

    expect(onUpdateInfo).toHaveBeenCalledTimes(1);
    expect(onUpdateInfo).toHaveBeenCalledWith('mock-id', {
      favorites: true,
    });
  });
});
