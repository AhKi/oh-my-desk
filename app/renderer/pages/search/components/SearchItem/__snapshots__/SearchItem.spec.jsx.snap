// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SearchItem Component should match to snapshot 1`] = `
<li
  className="SearchItem"
>
  <div
    className="SearchItem__toggle-btn"
  >
    <ToggleButton
      checkedValue={
        <i
          className="fa fa-check"
        />
      }
      isCheck={false}
      onToggle={[Function]}
      unCheckedValue={
        <i
          className="fa fa-times"
        />
      }
      width={40}
    />
  </div>
  <p
    className="SearchItem__Text"
  >
    <button
      className="SearchItem__Btn"
      onClick={[Function]}
      onDoubleClick={[Function]}
      type="button"
    >
      <strong
        className="SearchItem__title-text"
      >
        <span>
          mock-name
        </span>
      </strong>
      <span
        className="SearchItem__url-text"
      >
        <span>
          mock-url
        </span>
      </span>
    </button>
  </p>
  <p
    className="SearchItem__BtnSet"
  >
    <button
      className="SearchItem__bookmark-btn"
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        className="SearchItem__icon"
        src={Object {}}
      />
    </button>
  </p>
</li>
`;

exports[`Test SearchItem Component should match to snapshot when item.searched === both 1`] = `
<li
  className="SearchItem"
>
  <div
    className="SearchItem__toggle-btn"
  >
    <ToggleButton
      checkedValue={
        <i
          className="fa fa-check"
        />
      }
      isCheck={false}
      onToggle={[Function]}
      unCheckedValue={
        <i
          className="fa fa-times"
        />
      }
      width={40}
    />
  </div>
  <p
    className="SearchItem__Text"
  >
    <button
      className="SearchItem__Btn"
      onClick={[Function]}
      onDoubleClick={[Function]}
      type="button"
    >
      <strong
        className="SearchItem__title-text"
      >
        <HighlightParagraph
          content="mock-name"
          highlightClass=""
          keyword=""
        />
      </strong>
      <span
        className="SearchItem__url-text"
      >
        <HighlightParagraph
          content="mock-url"
          highlightClass=""
          keyword=""
        />
      </span>
    </button>
  </p>
  <p
    className="SearchItem__BtnSet"
  >
    <button
      className="SearchItem__bookmark-btn"
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        className="SearchItem__icon"
        src={Object {}}
      />
    </button>
  </p>
</li>
`;
